File: backend/routes/reports.js
import express from 'express';
import { authMiddleware } from '../middleware/auth.js';
import { monthIncomeExpense, spendByCategory } from '../utils/db.js';

const router = express.Router();
router.use(authMiddleware);

const monthSequence = (n) => {
  const today = new Date();
  const months = [];
  let y = today.getFullYear();
  let m = today.getMonth() + 1;
  
  for (let i = 0; i < n; i++) {
    months.push(`${y}-${String(m).padStart(2, '0')}`);
    m--;
    if (m === 0) {
      m = 12;
      y--;
    }
  }
  
  return months.reverse();
};

router.get('/', async (req, res) => {
  try {
    const userId = req.user.id;
    const months = monthSequence(6);
    
    // Income/Expense per month for last 6 months
    const month_ie = await monthIncomeExpense(userId, months);
    
    // Category breakdown for current month (expenses)
    const curr_month = months[months.length - 1];
    const spend_map = await spendByCategory(userId, curr_month);
    
    const cat_rows = Object.entries(spend_map)
      .map(([category, total]) => ({ category, total }))
      .sort((a, b) => b.total - a.total);
    
    res.json({
      months,
      month_income: months.map(m => month_ie[m]?.income || 0),
      month_expense: months.map(m => month_ie[m]?.expense || 0),
      categories: cat_rows.map(r => r.category),
      cat_totals: cat_rows.map(r => r.total),
      curr_month,
    });
  } catch (error) {
    res.status(500).json({ message: 'Server error' });
  }
});

export default router;

File: reports.py
from flask import Blueprint, render_template, g
from datetime import date, timedelta
from collections import defaultdict
from auth import login_required
from db import month_income_expense, spend_by_category

bp = Blueprint('reports', __name__, url_prefix='/reports')


def month_sequence(n: int):
    today = date.today().replace(day=1)
    months = []
    y, m = today.year, today.month
    for _ in range(n):
        months.append(f"{y:04d}-{m:02d}")
        m -= 1
        if m == 0:
            m = 12
            y -= 1
    return list(reversed(months))


@bp.route('/')
@login_required
def index():
    user = g.user
    months = month_sequence(6)

    # Income/Expense per month for last 6 months (MongoDB)
    month_ie = month_income_expense(user['id'], months)

    # Category breakdown for current month (expenses)
    curr_month = months[-1]
    spend_map = spend_by_category(user['id'], curr_month)
    cat_rows = sorted([
        {'category': k, 'total': v} for k, v in spend_map.items()
    ], key=lambda x: x['total'], reverse=True)

    return render_template(
        'reports/index.html',
        months=months,
        month_income=[month_ie[m]['income'] for m in months],
        month_expense=[month_ie[m]['expense'] for m in months],
        categories=[r['category'] for r in cat_rows],
        cat_totals=[float(r['total']) for r in cat_rows],
        curr_month=curr_month,
    )
